src/plugins/lightscript.js

```diff
@@ -65,64 +65,62 @@ pp.rewriteOperator = function (node) {
   }
 };
 
-pp.parseForFrom = function (node, iterator) {
-  this.expectContextual("from");
-  // `for i from`
-  const arrayOrRangeStart = this.parseExpression(true);
-  if (this.match(tt._thru) || this.match(tt._til)) {
-    return this.parseForFromRange(node, iterator, arrayOrRangeStart);
+function getIterableForKeyword(name) {
+  if (name === "idx" || name === "elem") {
+    return "array";
+  } else if (name === "key" || name === "val") {
+    return "object";
   } else {
-    return this.parseForFromArray(node, iterator, null, arrayOrRangeStart);
+    return null;
   }
-};
+}
 
-pp.parseForFromArray = function (node, iterator, element = null, array = null) {
-  if (iterator.type !== "Identifier") this.unexpected(iterator.start);
-  if (element && element.type !== "Identifier") this.unexpected(element.start);
-  node.id = iterator;
-  node.elem = element;
+pp.parseEnhancedForIn = function (node) {
+  const firstIterationType = this.state.value;
+  const iterableType = getIterableForKeyword(firstIterationType);
 
-  if (array) {
-    node.array = array;
-  } else {
-    this.expectContextual("from");
-    node.array = this.parseMaybeAssign(true);
-  }
+  this.next();
+  this.check(tt.name);
+  const firstIterationVar = this.state.value;
+  this.next();
 
-  this.expectParenFreeBlockStart();
-  node.body = this.parseStatement(false);
-  return this.finishNode(node, "ForFromArrayStatement");
-};
+  let secondIterationType = null, secondIterationVar = null;
+  if (this.match(tt.comma)) {
+    this.next();
+    this.check(tt.name);
+    secondIterationType = this.state.value;
+    if (!(iterableType === getIterableForKeyword(secondIterationType))) {
+      this.raise(this.state.lastTokStart, "Mismatched iteration keywords in for-in loop.");
+    }
 
-pp.parseForFromRange = function (node, iterator = null, rangeStart = null) {
-  if (iterator && iterator.type === "SequenceExpression") {
-    this.unexpected(iterator.expressions[1].start, "Ranges only iterate over one variable.");
+    this.next();
+    this.check(tt.name);
+    secondIterationVar = this.state.value;
+    this.next();
   }
-  node.id = iterator;
 
-  if (rangeStart) {
-    // `for 0 til`
-    if (rangeStart.type === "SequenceExpression") {
-      this.unexpected(rangeStart.expressions[1].start, "Unexpected comma.");
-    }
-    node.rangeStart = rangeStart;
-  } else {
-    // `for i from 0 til`
-    this.expectContextual("from");
-    node.rangeStart = this.parseMaybeAssign(true);
-  }
+  this.expect(tt._in);
 
-  if (this.eat(tt._thru)) {
-    node.inclusive = true;
+  if ( (firstIterationType === "idx") || (firstIterationType === "key") ) {
```

Comment:
these assignments can happen inline, above, eg; 

```js
if (this.eatContextual("idx")) {
  node.indexVariable = this.parseIdentifier();
} else if (this.eatContextual("key") {
  node.indexVariable = this.parseIdentifier();
} else if (this.eatContextual("val") {
  node.valueVariable = this.parseIdentifier();
} else if (this.eatContextual("elem") {
  node.valueVariable = this.parseIdentifier();
} else {
  this.unexpected();
}
```

This can be extracted into a `this.parseIterationVariable(node)` or similar function, which can be called twice: 

```js
this.parseIterationVariable(node);
if (this.eat(tt.comma) {
  this.parseIterationVariable(node);
  // ... check that they don't clash ...
}